platform/consul/controller.go

```diff
@@ -0,0 +1,219 @@
+
+	return data
+}
+
+func (c *Controller) getCatalogService(name string, q *api.QueryOptions) []*api.CatalogService {
+	endpoints, _, err := c.client.Catalog().Service(name, "", q)
+	if err != nil {
+		glog.Warningf("Could not retrieve service catalogue from consul: %v", err)
+		return []*api.CatalogService{}
+	}
+
+	return endpoints
+}
+
+// Instances retrieves instances for a service and its ports that match
+// any of the supplied tags. All instances match an empty tag list.
+func (c *Controller) Instances(hostname string, ports []string, tags model.TagsList) []*model.ServiceInstance {
+	// Get actual service by name
+	name, err := parseHostname(hostname)
+	if err != nil {
+		glog.V(2).Infof("parseHostname(%s) => error %v", hostname, err)
+		return nil
+	}
+
+	endpoints := c.getCatalogService(name, nil)
+
+	instances := []*model.ServiceInstance{}
+	for _, endpoint := range endpoints {
+		inst := convertInstance(endpoint)
+		if tags.HasSubsetOf(inst.Tags) && portMatch(inst, ports) {
+			instances = append(instances, inst)
+		}
+	}
+
+	return instances
+}
+
+// returns true if an instance's port matches with any in the provided list
+func portMatch(inst *model.ServiceInstance, ports []string) bool {
+	if len(ports) == 0 {
+		return true
+	}
+
+	for _, port := range ports {
+		if inst.Endpoint.ServicePort.Name == port {
+			return true
+		}
+	}
+
+	return false
+}
+
+// HostInstances lists service instances for a given set of IPv4 addresses.
+func (c *Controller) HostInstances(addrs map[string]bool) []*model.ServiceInstance {
+	data := c.getServices()
+	out := make([]*model.ServiceInstance, 0)
+	for svcName := range data {
+		for addr := range addrs {
+			endpoints := c.getCatalogService(svcName, nil)
+			for _, endpoint := range endpoints {
+				if addr == endpoint.ServiceAddress {
+					out = append(out, convertInstance(endpoint))
+				}
+			}
+		}
+	}
+
+	return out
+}
+
+// Run all controllers until a signal is received
+func (c *Controller) Run(stop <-chan struct{}) {
+	c.monitor.Start(stop)
+}
+
+// GetIstioServiceAccounts implements model.ServiceAccounts interface TODO
+func (c *Controller) GetIstioServiceAccounts(hostname string, ports []string) []string {
+	return nil
+}
+
+// AppendServiceHandler implements a service catalog operation
+func (c *Controller) AppendServiceHandler(f func(*model.Service, model.Event)) error {
+	c.monitor.AppendServiceHandler(func(obj interface{}, event model.Event) error {
+		f(convertService(*obj.(*[]*api.CatalogService)), event)
+		return nil
+	})
+	return nil
+}
+
+// AppendInstanceHandler implements a service catalog operation
+func (c *Controller) AppendInstanceHandler(f func(*model.ServiceInstance, model.Event)) error {
+	c.monitor.AppendInstanceHandler(func(obj interface{}, event model.Event) error {
+		f(convertInstance(&(*obj.(*api.CatalogService))), event)
+		return nil
+	})
+	return nil
+}
+
+// HasSynced implements model.ConfigStoreCache operation TODO
+func (c *Controller) HasSynced() bool {
```

Comment:
I'd suggest removing ConfigStore/Cache out of this PR. It's not necessary for service discovery, and is a separate interface.